File: L11/pi/pi.f
      program compute_pi
        implicit none
        include "omp_lib.h"
        
        integer n, i, numthr
        character*32 inarg
        parameter (n = 9100)
        double precision w, x, sum, pi, f, y, piref
        parameter (piref = 3.141592653589793238d0)
        
        f(y) = 4.d0/(1.d0+y*y)
  
        call get_command_argument(1, inarg)
        read(inarg,'(i2)') numthr

        call omp_set_num_threads(numthr)

        print'(a,i2,a,i2,a)', 
     &          'The paralel part of program is running on ',
     &          omp_get_max_threads(), ' threads, ',
     &          omp_get_num_procs(), ' processors'

        w = 1.0d0/n
        sum = 0.0d0;
!$omp parallel do private(x) shared(w)
!$omp& reduction(+:sum)
        do i=1,n
          x = w*(i-0.5d0)
          sum = sum + f(x)
        end do
        
        pi = w*sum
        print'(a, f18.16)', 'pi    = ', pi
        print'(a, f18.16)', 'piref = ', piref
        print'(a, e9.3)', 'The difference is ', abs(pi-piref)
      end

File: L11/pi/pi.c
#include <stdio.h>
#include <omp.h>
#include <math.h>

const double piRef = 3.141592653589793238;

double f(double y) {return(4.0/(1.0+y*y));}

int main(int argc, const char** argv)
{
    int num_threads;
    sscanf(argv[1], "%d", &num_threads);

    omp_set_num_threads(num_threads);

    double w, x, sum, pi;
    int i;
    int n = 9100;
    w = 1.0/n;
    sum = 0.0;

    double test = 1.1;
 
    printf("The paralel part of program is running on %d"\
           " threads, %d processors \n",
            omp_get_max_threads(), omp_get_num_procs());

    #pragma omp parallel for private(x) shared(w)\
    reduction(+:sum)
    for(i=0; i < n; i++)
    {
        x = w*(i+0.5);
        sum = sum + f(x);
    }

    pi = w*sum;
    printf("pi    = %.16f\n", pi);
    printf("piRef = %.16f\n", piRef);
    printf("The differnce is %.2e \n", fabs(pi - piRef));

    return 0;
}